API/common/service_impl/chatgpt.py

```diff
@@ -23,29 +24,76 @@
 openai.api_base = os.environ["OPENAI_API_URL"]
 
 
-@timing_decorator
-def gpt_generate(system_msg: str, user_msg: str, history_msg: List[Dict[str, str]]) -> Dict[str, Any]:
+@log_dependency_call("gpt generate")
+def gpt_generate(context, system_msg: str, user_msg: str, history_msg: List[Dict[str, str]]) -> Dict[str, Any]:
     # the param dict of the chatgpt service
     chatgpt_service_params = initialize_chatgpt_service_params(system_msg)
     all_user_msg = []
+    estimated_history_tokens = 0
     if history_msg:
+        estimated_history_tokens = num_tokens_from_messages(history_msg)
         chatgpt_service_params["messages"].extend(history_msg)
         all_user_msg = [msg["content"]
                         for msg in history_msg if msg["role"] == "user"]
     all_user_msg.append(user_msg)
     chatgpt_service_params["messages"].append(
         {"role": "user", "content": "\n".join(all_user_msg)})
+    
+    # get the estimated number of tokens
+    estimated_prompt_tokens = num_tokens_from_messages(chatgpt_service_params["messages"])
+    gpt_task_name = context.custom_context.gpt_task_name
+    del context.custom_context.gpt_task_name
+    estimated_question_tokens = context.custom_context.estimated_question_tokens
+    del context.custom_context.estimated_question_tokens
+    estimated_task_list_tokens = getattr(context.custom_context, 'estimated_task_list_tokens', None)
+    if estimated_task_list_tokens:
+        del context.custom_context.estimated_task_list_tokens
+    task_list_lens = getattr(context.custom_context, 'task_list_lens', None)
+    if task_list_lens:
+        del context.custom_context.task_list_lens
+    estimated_usage_context_tokens = getattr(context.custom_context, 'estimated_usage_context_tokens', None)
+    if estimated_usage_context_tokens:
+        del context.custom_context.estimated_usage_context_tokens
+
+    # logging GPT call cost
+    message = f"The estimated cost of {gpt_task_name} GPT call is as follows: "
+    message += f"question tokens = {estimated_question_tokens}, "
+    message += f"task list tokens = {estimated_task_list_tokens}, task lens = {task_list_lens}, " if estimated_task_list_tokens else ""
+    message += f"usage context tokens = {estimated_usage_context_tokens}, " if estimated_usage_context_tokens else ""
+    message += f"history tokens = {estimated_history_tokens}, "
+    message += f"propmpt tokens = {estimated_prompt_tokens}."
+    logging.info(f"{message}")
```

Comment:
Could we abstract these logic into a separate method and call it here? 
Since there is also a lot of logic for logging, it may be better to separate it from the business processing logic